### ISSUE TITLE
Support Turing's SimpleVarInfo

### ISSUE THREAD

**Issue by yebai** [withbayes/Tapir.jl#132]
I run into the following issue while playing with Tapir on big Turing models.

```julia
julia> using Turing, AbstractMCMC, Tapir, ADTypes, LogDensityProblems, LogDensityProblemsAD


julia> function AbstractMCMC.LogDensityModel(m::Turing.DynamicPPL.Model, adtype::ADTypes.AbstractADType)
       f = LogDensityFunction(m, DynamicPPL.SimpleVarInfo(m))
       adf = AbstractMCMC.LogDensityModel(LogDensityProblemsAD.ADgradient(adtype, f))
       return adf
       end

julia> num_tildes = 50 # setting this to 5 runs fine. 
50

julia> expr = :(function $(Symbol(:demo, num_tildes))() end) |> Base.remove_linenums!;

julia> mainbody = last(expr.args);

julia> append!(mainbody.args, [:($(Symbol("x", j)) ~ Normal()) for j = 1:num_tildes]);

julia> f = @eval $(DynamicPPL.model(:Main, LineNumberNode(1), expr, false))
demo50 (generic function with 2 methods)

julia> model = f();

julia> f = AbstractMCMC.LogDensityModel(model, AutoTapir());

julia> initial_params = rand(LogDensityProblems.dimension(f.logdensity));

julia> LogDensityProblems.logdensity_and_gradient(f.logdensity, initial_params)
```

**Comment by willtebbutt** [comment]
Oh, interesting. I guess I just haven't previously encountered a `splatnew` expression. This definitely needs to be supported, and should only be a small extension of what we currently have. I'll try to isolate a MWE.

**Comment by willtebbutt** [comment]
Anyway, I think I basically understand what appears to be going on here. To summarise:
1. Tapir doesn't support :splatnew (I'm pleased with the error message that was thrown -- it was trivial to debug), but it should.
2. Tapir doesn't support `Core._apply_iterate` but it should. We already knew this (see the note on the readme).
3. SimpleVarInfo has inference issues for large numbers of tilde statements. I don't know if this was known, but it's very solvable -- see the issue I've opened https://github.com/TuringLang/DynamicPPL.jl/issues/600 .

Additionally, it may transpire that there are more problems uncovered by solving these three problems.

If we're very keen to make SimpleVarInfo type stable for moderate numbers of tilde statments then I won't make solving 1 and 2 incredibly high priority -- they definitely need to be resolved at some point though.

**Comment by willtebbutt** [comment]
splatnew added in #141 -- will turn my attention to apply_iterate this week.

**Comment by willtebbutt** [comment]
#152 makes `Core._apply_iterate` work, and improves a couple of misc rrules. Doesn't look like it's everything we need to make Tapir work with Turing's VarInfo, so I'm expanding the scope of this issue to that.

**Comment by willtebbutt** [comment]
This issue is resolved via #160 , specifically for SimpleVarInfo. If you encounter any more issues with SimpleVarInfo, please re-open. Note that the performance remains not great, but it does at least run.
